### PR TITLE
Implement trait based API for defining WindowUDF

### LINKED ISSUE #8711: Implement trait based API for defining `WindowUDF`
Similarly to https://github.com/apache/arrow-datafusion/issues/8568

### Is your feature request related to a problem or challenge?

The current way a user implements a [WindowUDF](https://docs.rs/datafusion/latest/datafusion/logical_expr/struct.WindowUDF.html#) is awkward and very hard to extend in backwards compatible ways:

They must wade through several `Arc<dyn<...>` typedefs to figure out how to provide the type signature and implementation

```rust
impl WindowUDF {

/// Create a new WindowUDF 
pub fn new(
    name: &str,
    signature: &Signature,
    return_type: &Arc<dyn Fn(&[DataType]) -> Result<Arc<DataType>, DataFusionError> + Send + Sync>,
    partition_evaluator_factory: &Arc<dyn Fn() -> Result<Box<dyn PartitionEvaluator>, DataFusionError> + Send + Sync>
) -> WindowUDF
```

### Describe the solution you'd like

Follow the pattern in https://github.com/apache/arrow-datafusion/pull/8578

1. Create a `WindowUDFImpl` trait, and `WindowUDF::new_from_impl` that creates an `WindowUDF` from the impl
2. Add an example in `datafusion-examples/examples/advanced_udwf.rs` of using this API

I am not sure why this API is implemented like it is (other than it was consistent with `ScalarUDF`). As a user I would expect to be able to use a trait object like this

like

```rust
struct MyWindowFunction { 
..
}

impl WindowUDFImpl for MyWindowFunction {
  fn name(&self) -> &str, 
  fn return_type(&self) -> &DataType, 
...
}
```

### Describe alternatives you've considered

_No response_

### PR DESCRIPTION
**Pull request by guojidan**
## Which issue does this PR close?

Closes #8711 .

### CODE REVIEW

**Comment by guojidan** [comment]
cc @alamb

**Comment by alamb** [comment]
Thank you @guojidan  -- I plan to review this carefully later today

File: datafusion/expr/src/udwf.rs [hunk 1]
```diff
@@ -140,3 +171,86 @@ impl WindowUDF {
...
+    /// Returns the function's [`Signature`] for information about what input
+    /// types are accepted and the function's Volatility.
+    fn signature(&self) -> &Signature;
+
+    /// What [`DataType`] will be returned by this function, given the types of
+    /// the arguments
+    fn return_type(&self, arg_types: &[DataType]) -> Result<DataType>;
+
+    /// Invoke the function, returning the [`PartitionEvaluator`] instance
+    fn invoke(&self) -> Result<Box<dyn PartitionEvaluator>>;
```
Comment (alamb) [inline datafusion/expr/src/udwf.rs:255]: Would it be possible to rename this function to `partition_evaluator` as it isn't really being invoked?

```suggestion
    fn partition_evaluator(&self) -> Result<Box<dyn PartitionEvaluator>>;
```

File: datafusion-examples/examples/advanced_udwf.rs [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+};
+use datafusion::error::Result;
+use datafusion::prelude::*;
+use datafusion_common::ScalarValue;
+use datafusion_expr::{
+    PartitionEvaluator, Signature, WindowFrame, WindowUDF, WindowUDFImpl,
+};
+
+/// This example shows how to use the full WindowUDFImpl API to implement a user
+/// defined function. As in the `simple_udwf.rs` example, this struct implements
```
Comment (alamb) [inline datafusion-examples/examples/advanced_udwf.rs:33]:
```suggestion
/// defined window function. As in the `simple_udwf.rs` example, this struct implements
```

**Review by alamb — APPROVED** [review]
Wow @guojidan  -- 👏  this is a really nice inital PR 

I had one very small suggestion related to naming

I'll also file follow on tickets for similar cleanups as we did for ScalarFunctions (namely implement `WindowUDF` in terms of the trait)

But again, thank you for this PR

**Comment by guojidan** [comment]
@alamb Thank you for your affirmation, it gives me confidence to continue contributing this repository : )

I have modified the code according to your suggestion.

**Review by alamb — APPROVED** [review]

**Comment by alamb** [comment]
Filed https://github.com/apache/arrow-datafusion/issues/8734 as a follow on

**Pull request merged**